pkg/mic/mic.go

```diff
@@ -401,57 +441,166 @@ func (c *Client) makeAssignedIDs(azID *aadpodid.AzureIdentity, azBinding *aadpod
 	return assignedID, nil
 }
 
-func (c *Client) createAssignedIdentityDeps(assignedID *aadpodid.AzureAssignedIdentity, id *aadpodid.AzureIdentity, node *corev1.Node) error {
+func (c *Client) createAssignedIdentity(assignedID *aadpodid.AzureAssignedIdentity) error {
 	err := c.CRDClient.CreateAssignedIdentity(assignedID)
 	if err != nil {
 		glog.Error(err)
 		return err
 	}
+	return nil
+}
+
+func (c *Client) removeAssignedIdentity(assignedID *aadpodid.AzureAssignedIdentity) error {
+	err := c.CRDClient.RemoveAssignedIdentity(assignedID)
+	if err != nil {
+		glog.Error(err)
+	}
+	return nil
+}
+
+func (c *Client) appendToRemoveListForNode(nodeMap map[string]trackUserAssignedMSIIds, resourceID string, node *corev1.Node) {
+	if trackList, ok := nodeMap[node.Name]; ok {
+		if trackList.removeUserAssignedMSIIDs != nil {
+			trackList.removeUserAssignedMSIIDs = append(trackList.removeUserAssignedMSIIDs, resourceID)
+			trackList.node = node
+		} else {
+			trackList.removeUserAssignedMSIIDs = []string{resourceID}
+			trackList.node = node
+		}
+		nodeMap[node.Name] = trackList
+	} else {
+		nodeMap[node.Name] = trackUserAssignedMSIIds{removeUserAssignedMSIIDs: []string{resourceID}, node: node}
+	}
+}
+
+func (c *Client) appendToAddListForNode(nodeMap map[string]trackUserAssignedMSIIds, resourceID string, node *corev1.Node) {
+	if trackList, ok := nodeMap[node.Name]; ok {
+		if trackList.addUserAssignedMSIIDs != nil {
+			trackList.addUserAssignedMSIIDs = append(trackList.addUserAssignedMSIIDs, resourceID)
+			trackList.node = node
+		} else {
+			trackList.addUserAssignedMSIIDs = []string{resourceID}
+			trackList.node = node
+		}
+		nodeMap[node.Name] = trackList
+	} else {
+		nodeMap[node.Name] = trackUserAssignedMSIIds{addUserAssignedMSIIDs: []string{resourceID}, node: node}
+	}
+}
 
-	if id.Spec.Type == aadpodid.UserAssignedMSI {
+func (c *Client) handleNodeErrors(nodesWithError []string, addList *[]aadpodid.AzureAssignedIdentity, deleteList *[]aadpodid.AzureAssignedIdentity, newAssignedIDs []aadpodid.AzureAssignedIdentity) error {
 
-		err = c.CloudClient.AssignUserMSI(id.Spec.ResourceID, node)
+	nodeIdentityList := make(map[string][]string)
+	for _, n := range nodesWithError {
+		// get the list of msi's currently on node
+		node, err := c.NodeClient.Get(n)
 		if err != nil {
-			glog.Error(err)
-			newErr := c.CRDClient.RemoveAssignedIdentity(assignedID)
-			if newErr != nil {
-				glog.Errorf("Error when removing assigned identity in create error path err: %v", newErr)
+			glog.Errorf("Lookup of node %s resulted in error %v", n, err)
+			continue
+		}
+		idList, err := c.getUserMSIListForNode(node)
+		if err != nil {
```

Comment:
What is the plan here. If we just log and continue then idList can be having random values..?

Reply:
getUserMSIListForNode returns nil on error. So we'll fail all of them.